torch/csrc/autograd/engine.cpp

```diff
@@ -368,6 +382,23 @@ auto Engine::evaluate_function(FunctionTask& task) -> void {
     if (!fn_info.needed) return;
   }
 
+  // Sets stream (and gpu) for the function to run on
+  #ifdef USE_CUDA 
+    int parent_device = -1;
+    THCStream* parent_stream = nullptr;
+    for (auto i = decltype(task.inputs.size()){0}; i < task.inputs.size(); ++i) {
+      auto& var = task.inputs[i];
+      if (var.defined() && var.type().is_cuda()) {
+        parent_device = var.get_device();
+        parent_stream = task.fn->input_metadata(i).stream();
+        break;
+      }      
+    }
```

Comment:
1. There's already `input_buffer.device()`, so it would be better to use this, or extend it with a possibility of returning a stream.
2. I think AutoGPU is unnecessary here, since you're already guaranteed to be on the thread associated with the device where the inputs live.

Reply:
You are correct that we could AutoStream here instead of AutoGPUStream. Might as well.

Extending input_buffer.device() sounds find and should make the code look cleaner, anyway.